File: leetcode/deepest_leaves_sum.py
#!/usr/bin/env python3

from collections import deque
import sys
import unittest

# number: 1302
# title: Deepest Leaves Sum
# url: https://leetcode.com/problems/deepest-leaves-sum/
# section:
# difficulty: medium
# tags: tree, dfs, bfs, binary tree

# constraints
# The number of nodes in the tree is in the range [1, 10^4].
# 1 <= Node.val <= 100

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# solution: iterative bfs
# complexity
# run-time: O(n)
# space: O(n)
def deepest_leaves_sum(root):
    queue = deque([root])

    while queue:
        current_len = len(queue)
        # reset sum on every level
        ans = 0

        for _ in range(current_len):
            node = queue.popleft()
            # count sum on every level
            ans += node.val

            if node.left:
                queue.append(node.left)

            if node.right:
                queue.append(node.right)

    # sum on deepest level will not be reset
    return ans

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/deepest_leaves_sum.cpp
#include <queue>

// number: 1302
// section:
// difficulty: medium
// tags: tree, dfs, bfs, binary tree

// constraints
// The number of nodes in the tree is in the range [1, 10^4].
// 1 <= Node.val <= 100

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
    : val(x), left(left), right(right) {}
};

// complexity
// run-time: O(n)
// space: O(n)
int deepestLeavesSum(TreeNode* root)
{
    std::queue<TreeNode*> queue;
    queue.push(root);
    int ans = 0;

    while (!queue.empty()) {
        int len = queue.size();
        ans = 0;

        for (size_t i = 0; i != len; ++i) {
            TreeNode* node_p = queue.front();
            queue.pop();

            ans += node_p->val;

            if (node_p->right) {
                queue.push(node_p->right);
            }

            if (node_p->left) {
                queue.push(node_p->left);
            }
        }
    }

    return ans;
}
